### ISSUE TITLE
[Vanity URLs] : Changes to "Site" field are not being persisted

### ISSUE THREAD

**Issue by jcastro-dotcms** [dotCMS/core#21373]
Found when working on Support Ticket: [105909](https://dotcms.zendesk.com/agent/tickets/105909)
Reproducible as of dotCMS 5.3.8.6 up to latest master (21.11)

Vanity URLs can work for either a specific Site or all Sites in a dotCMS instance. If you change where a given Vanity URL is supposed to work on, it should work as expected. However, when doing so, changes are not getting picked up and the Vanity URL still points to the previous  value -- either a specific Site, or all Sites.

Flushing the Cache Regions and/or doing a full reindex is not fixing the problem.

### NOTE TO DEVS #2: At least in dotCMS 5.3.8.6, adding the `Site or Folder` field does not solve the problem. The back-end is ignoring any change you make to it and the new Site you select is not persisted.

**To Reproduce**
1. Create a copy of the `demo.dotcms.com` Site named `second.dotcms.com`.
2. Select the `demo.dotcms.com` Site, and create a Vanity URL with the following information. **It's very important the the Vanity points to `All sites` first**:
2.1. Title: `My test Vanity`
2.2. Site: `All sites`
2.3. Uri: `/my-test`
2.4. Action: `301 - Redirect`
2.5. Forward To: `/index`
3. The Vanity URL works for `demo.dotcms.com`:
```
$ curl -sk --head -H "Host: demo.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:42 GMT
Connection: close
```
4. And for `second.dotcms.com` as well, which is expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:58 GMT
Connection: close
```
5. Now, change the Vanity URL to point to `demo.dotcms.com` **ONLY**, and try both CURL commands again.
6. Both of them are returning Status Code 301, which is **NOT** correct as the one sent to `second.dotcms.com` should have failed with a 404.
7. If you re-create the Vanity URL and point it to the `demo.dotcms.com` Site since the beginning, the CURL command to `demo.dotcms.com` works again, and the one to `second.dotcms.com` fails as expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 404 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Set-Cookie: JSESSIONID=D6BA571A1954E4ABD0E9A5951245AEFE; Path=/; Secure; HttpOnly; SameSite=Strict
Content-Type: text/html;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:45:09 GMT
Connection: close
```

**Expected behavior**
Changing the **Site** field must update the information of the Vanity URL correctly.

**Additional context**
The only workaround is to re-create the Vanity URL from scratch. However, this is not a feasible solution for the customer. We need to look at the possibility of creating a SQL query that could update the incorrect and fix this automatically.

In the customer instance, running 5.3.8.6, I was able to confirm that the "more technical" workaround for fixing a Vanity URL that is not working is the following:

- Look for the Vanity URL in the `identifier` table, and update the `host_inode` column to point to either the right site or `SYSTEM_HOST` if you need it to apply to all Sites.
- Flush the region called `Vanity URL Cache`.
- Test the Vanity URL once more. You can use CURL, like this:
```
curl -sk --head -H "Host: www.yoursite.com" https://your-dotcms-instance.com/your-vanity-url
```

**Comment by wezell** [comment]
This is because the "site" field is a fake host/folder field and gets out of sync with the host/folder field. We need to update the site field to be a real host/folder field instead of a custom field and make that work.

**Comment by jcastro-dotcms** [comment]
The currently valid workaround is to use the following OSGi Plugin: https://github.com/dotcms-plugins/com.dotcms.move-actionlet

This plugin adds a Custom Workflow Sub-action named `Dot Move` which allows content authors to move the conflicting Vanity URL to the correct Site. Such a sub-action provides a single text field where users enter the name of the Site they need to move the Vanity URL to. For example:
```
//www.yoursite.com/
```
The value is made up of:
1. The initial double slashes.
2. The Site's name.
3. The trailing slash.

**Comment by fabrizzio-dotCMS** [comment]
I tried this against the latest master and for the most part.  I think it worked. 
Here's what I did.

I created a vanity having it return a 301 to  /contact-us/index and made it point to all sites.
Tried the command and the Vanity  worked for both sites 

`curl -sk --head -H "Host: demo.dotcms.com"  http://127.0.0.1:8080/lol`
`curl -sk --head -H "Host: demo2.dotcms.com"  http://127.0.0.1:8080/lol`

I got a 301 response code for both

Curiously I still get a 301 even if I send a Host head with any random string

Then I changed the Vanity to point to the site `demo.dotcms.com` only hitting save. 

Verified the db using the new inode and I see the changes had taken effect. 

Tried the first culr again and still got a 301
Then I hit publish on the UI and then it all worked for me. 
Seems the changes take effect until they get published. At least in the current master.
